src/pcapng/blocks/custom.rs

```
@@ -0,0 +1,131 @@
+//! Custom Block.
+
+use std::borrow::Cow;
+use std::io::Write;
+
+use byteorder_slice::ByteOrder;
+use byteorder_slice::byteorder::{ReadBytesExt, WriteBytesExt};
+
+use super::block_common::{Block, PcapNgBlock};
+use crate::pcapng::PcapNgState;
+use crate::PcapError;
+
+/// Custom block
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CustomBlock<'a, const COPIABLE: bool> {
+    /// Private Enterprise Number of the entity which defined this block.
+    pub pen: u32,
+    /// Payload of this block.
+    pub payload: Cow<'a, [u8]>,
+}
+
+impl<'a> CustomBlock<'a, true> {
+    /// Converts this block's payload into a type that implements [`CustomCopiable`].
+    pub fn interpret<T: CustomCopiable<'a>>(&'a self) -> Result<Option<T>, PcapError> {
+        if self.pen != T::PEN {
+            return Ok(None)
+        }
+
+        T::from_slice(&self.payload)
+    }
+}
+
+impl<'a> CustomBlock<'a, false> {
+    /// Converts this block's payload into a type that implements [`CustomNonCopiable`].
+    pub fn interpret<T: CustomNonCopiable<'a>>(&'a self, state: &T::State)
+        -> Result<Option<T>, PcapError>
+    {
+        if self.pen != T::PEN {
+            return Ok(None)
+        }
+
+        T::from_slice(state, &self.payload)
+    }
+}
+
+impl<const COPIABLE: bool> CustomBlock<'_, COPIABLE> {
+    // The into_owned method must be implemented manually,
+    // since derive_into_owned can't handle the const generic.
+
+    /// Returns a version of self with all fields converted to owning versions.
+    pub fn into_owned(self) -> CustomBlock<'static, COPIABLE> {
+        CustomBlock {
+            pen: self.pen,
+            payload: Cow::Owned(self.payload.into_owned())
+        }
+    }
+}
+
+/// Common interface for copiable custom block payloads
+pub trait CustomCopiable<'a> {
```

With `from_slice()` and `write_to() `the user may need to return custom errors. To support this use-case we could add an error type to both traits. What do you think ? 
In this case we would have to add a new variant to `PcapError` to handle the conversion between the user defined error and `PcapError` when converting a `CustomBlock` into a `Block`

---

> With `from_slice()` and `write_to() `the user may need to return custom errors. To support this use-case we could add an error type to both traits. What do you think ?

Yes this is a good point, limiting the user to only the current `PcapError` variants is probably too restrictive here. I'm not sure if adding an error type to the trait is the best way forward though.

Unless we add _two_ error types, then both_ `from_slice` and `write_to` would need to return the same error type. Which may be annoying in practice, since `from_slice` may fail in complex format-dependent ways, but `write_to` is likely to be just calling into methods from `WriteBytesExt`.

> In this case we would have to add a new variant to `PcapError` to handle the conversion between the user defined error and `PcapError` when converting a `CustomBlock` into a `Block`

If we make that variant something like `CustomConversionError(Box<dyn std::error::Error>)`, then we wouldn't need the error type in the trait, the `from_slice` and `write_to` methods would be able to return it directly.

---

Everything else seems fine to me. Once we've settled on this topic I'll approve the PR.